src/docker-utils.coffee

```diff
@@ -62,46 +62,68 @@ do ->
-					.on 'progress', (progress) ->
-						# In request-progress ^2.0.1, "percentage" is a ratio from 0 to 1
-						onProgress(percentage: progress.percentage * 100)
-					.on 'end', ->
-						onProgress(percentage: 100)
-					.on 'response', (res) ->
-						if res.statusCode is 504
-							reject(new Error('Delta server is still processing the delta, will retry'))
-						else if res.statusCode isnt 200
-							reject(new Error("Got #{res.statusCode} when requesting image from delta server."))
-						else
-							if imgSrc is 'resin/scratch'
-								deltaSrc = null
-							else
-								deltaSrc = imgSrc
-							res.pipe(dockerDelta.applyDelta(deltaSrc, imgDest))
-							.on('id', resolve)
-							.on('error', reject)
-					.on 'error', reject
+					docker.getRegistryAndName(app.imageId)
+					.then ({ registry, imageName, tagName }) ->
+						tokenEndpoint = "#{app.apiEndpoint}/auth/v1/token" #TODO read from state endpoint
+
+						request.get "#{tokenEndpoint}?service=#{registry}&scope=repository:#{imageName}:pull", authOpts
+						.spread (response, body) ->
+							opts =
+								timeout: requestTimeout
+							if body?.token?
+								deltaAuthOpts =
+									auth:
+										bearer: body?.token
+								opts = _.merge(opts, deltaAuthOpts)
+
+							progress request.get("#{config.deltaHost}/api/v2/delta?src=#{imgSrc}&dest=#{imgDest}", opts)
+							.on 'progress', (progress) ->
+								# In request-progress ^2.0.1, "percentage" is a ratio from 0 to 1
+								onProgress(percentage: progress.percentage * 100)
+							.on 'end', ->
+								onProgress(percentage: 100)
+							.on 'response', (res) ->
+								if res.statusCode is 504
+									reject(new Error('Delta server is still processing the delta, will retry'))
+								else if res.statusCode isnt 200
+									reject(new Error("Got #{res.statusCode} when requesting image from delta server."))
+								else
+									if imgSrc is 'resin/scratch'
+										deltaSrc = null
+									else
+										deltaSrc = imgSrc
+									res.pipe(dockerDelta.applyDelta(deltaSrc, imgDest))
+									.on('id', resolve)
+									.on('error', reject)
+							.on 'error', reject
 				.timeout(totalTimeout)
 			.then (id) ->
 				getRepoAndTag(imgDest)
 				.then ({ repo, tag }) ->
 					docker.getImage(id).tagAsync({ repo, tag, force: true })
 			.catch dockerDelta.OutOfSyncError, (err) ->
 				console.log('Falling back to delta-from-empty')
-				exports.rsyncImageWithProgress(imgDest, { requestTimeout, totalTimeout, startFromEmpty: true }, onProgress)
+				exports.rsyncImageWithProgress(imgDest, { requestTimeout, totalTimeout, authOpts, startFromEmpty: true }, onProgress)
 
 	exports.fetchImageWithProgress = (image, onProgress) ->
 		Promise.using readLockImages(), ->
-			dockerProgress.pull(image, onProgress)
+			Promise.join utils.getConfig('apiKey'), utils.getConfig('uuid'), (apiKey, uuid) ->
```

Comment:
I think this should be done in `application.fetch` (you already do it in the deltas case, you could just do it always before checking for the RESIN_SUPERVISOR_DELTA config). This makes it more consistent: you always get the uuid and apikey there, and pass the credentials to either fetchImageWithProgress or rsyncImageWithProgress.